ddtrace/tracer/textmap.go

```diff
@@ -487,3 +501,309 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
+	// tilde (reserved, will represent 0x3D (equals) in the encoded tag value,
+	// and characters outside the ASCII range 0x20 to 0x7E.
+	// Equals character must be encoded with a tilde.
+	// Other disallowed characters must be replaced with the underscore.
+	valueRgx := regexp.MustCompile(",|;|:|[^\\x20-\\x7E]+")
+
+	// originRgx is used to sanitize the value of the datadog origin tag.
+	// Disallowed characters are comma (reserved as a list-member separator),
+	// semi-colon (reserved for separator between entries in the dd list-member),
+	// equals (reserved for list-member key-value separator).
+	// Disallowed characters must be replaced with the underscore.
+	originRgx := regexp.MustCompile(",|=|;|[^\\x20-\\x7E]+")
+
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("dd=s:%d", priority))
+	listLength := 1
+
+	if ctx.origin != "" {
+		b.WriteString(fmt.Sprintf(";o:%s",
+			originRgx.ReplaceAllString(ctx.origin, "_")))
+	}
+
+	for k, v := range ctx.trace.propagatingTags {
+		if !strings.HasPrefix(k, "_dd.p.") {
+			continue
+		}
+		// Datadog propagating tags must be appended to the tracestateHeader
+		// with the `t.` prefix. Tag value must have all `=` signs replaced with a tilde (`~`).
+		tag := fmt.Sprintf("t.%s:%s",
+			keyRgx.ReplaceAllString(k[len("_dd.p."):], "_"),
+			strings.ReplaceAll(valueRgx.ReplaceAllString(v, "_"), "=", "~"))
+		if b.Len()+len(tag) > 256 {
+			break
+		}
+		b.WriteString(";")
+		b.WriteString(tag)
+	}
+	// the old state is split by vendors, must be concatenated with a `,`
+	if len(oldState) == 0 {
+		return b.String()
+	}
+	for _, s := range strings.Split(strings.Trim(oldState, " \t"), ",") {
+		if strings.HasPrefix(s, "dd=") {
+			continue
+		}
+		listLength++
+		// if the resulting tracestateHeader exceeds 32 list-members,
+		// remove the rightmost list-member(s)
+		if listLength > 32 {
+			break
+		}
+		b.WriteString("," + strings.Trim(s, " \t"))
+	}
+	return b.String()
+}
+
+func (p *propagatorW3c) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
+	switch c := carrier.(type) {
+	case TextMapReader:
+		return p.extractTextMap(c)
+	default:
+		return nil, ErrInvalidCarrier
+	}
+}
+
+func (*propagatorW3c) extractTextMap(reader TextMapReader) (ddtrace.SpanContext, error) {
+	var parentHeader string
+	var stateHeader string
+	// to avoid parsing tracestate header(s) if traceparent is invalid
+	if err := reader.ForeachKey(func(k, v string) error {
+		key := strings.ToLower(k)
+		switch key {
+		case traceparentHeader:
+			if parentHeader != "" {
+				return ErrSpanContextCorrupted
+			}
+			parentHeader = v
+		case tracestateHeader:
+			stateHeader = v
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	var ctx spanContext
+	if err := parseTraceparent(&ctx, parentHeader); err != nil {
+		return nil, err
+	}
+	if err := parseTracestate(&ctx, stateHeader); err != nil {
+		return nil, err
+	}
+	// ctx.traceID might be nil due to higher order bits, eg. 1....00000000
```

Comment:
I will remove that check since I have it in the function. But for the context, not exactly, since the traceID might be 0 and still be valid if the its first 16 chars are non 0 and the rest are : 1111......0000000000000000